test/cpp/test_xla_sharding.cpp

```diff
@@ -56,12 +57,32 @@ TEST_F(XLAShardingTest, ShardTensor) {
   EXPECT_EQ(shards[0].sizes(), c10::ArrayRef<long>({4, 2, 4}));
   EXPECT_EQ(shards[7].sizes(), c10::ArrayRef<long>({4, 1, 4}));
 
-  // Replicated, all shards should be identical
+  // 3D tiled, the first dim is replicated and the last halved. The last shard
+  // size should be smaller in dim=1 because it's not evenly divisible.
+  xla::Array3D<int64_t> cube({{{0, 1}, {2, 3}, {4, 5}, {6, 7}}});
+  sharding = xla::HloSharding::Tile(cube).ToProto();
+  shards = ShardingUtil::ShardTensor(tensor, sharding, devices);
+  EXPECT_EQ(shards.size(), 8);
+  EXPECT_EQ(shards[0].sizes(), c10::ArrayRef<long>({8, 2, 2}));
+  EXPECT_EQ(shards[7].sizes(), c10::ArrayRef<long>({8, 1, 2}));
+
+  // Replicated, all shards should be identical.
   sharding = xla::HloSharding::Replicate().ToProto();
   shards = ShardingUtil::ShardTensor(tensor, sharding, devices);
   EXPECT_EQ(shards.size(), 8);
   EXPECT_EQ(shards[0].sizes(), c10::ArrayRef<long>({8, 7, 4}));
   EXPECT_EQ(shards[7].sizes(), c10::ArrayRef<long>({8, 7, 4}));
+
+  // 4D tiled, the first and second dims are replicated and the last halved. The
```

Comment:
@steventk-g I've also added the test case for 4-dimension tensor input. Follow these examples (e.g., cute and tesseract cases), and it should be straightforward to add a case for 5-dimension input (let's do that in a separate PR, but this is lower priority than virtual device work). Writing the test case should be quick, but try to spend some extra time studying `ShardingUtil::ShardTensor` -- but let's box our time no more than half a day.  cc @JackCaoG

Reply:
@steventk-g Please open a github issue or bug to track this so we don't forget :D

Reply:
Done https://github.com/pytorch/xla/issues/4093

Reply:
This task is blocked because there is no `Array5D` class in XLA. It only goes up to 4D right now.

Reply:
@yeounoh do you know if we really need 5d or it is a good to have?

Reply:
it's a good to have. The ask was up to 3d.

Reply:
@steventk-g I think we could use the generic Array to represent the 5d case, but a low prio for now. Let me look into it later.